### PR TITLE
DOC (string dtype): updated 'Working with text data' for str dtype in pandas 3.0

### LINKED ISSUE #60348: DOC (string dtype): update user guide page "Working with text data"
With the new default string dtype in pandas 3.0, we should update the documentation to properly reflect this. The main page about working with string data is https://pandas.pydata.org/pandas-docs/version/2.2/user_guide/text.html

This page currently mentions `object` dtype vs nullable `StringDtype`, some differences, and then shows most examples using `dtype="string"`.

We should update that page to reflect that there is now a default `"str"` dtype, add a historical note about `object` dtype being the default before pandas 3.0 (and that you can still encounter this, and then how to convert to str dtype; and refer to the upgrade guide), mention the differences of "str" vs "string" dtype, etc

### PR DESCRIPTION
**Pull request by Uvi-12**
Closes #60348 

This PR updates the "Working with Text Data" page in the pandas documentation to reflect the change in pandas 3.0 where "str" dtype is now the default.

### CODE REVIEW

**Comment by Uvi-12** [comment]
Hi @mroeschke, Please review the PR. I've made some updates to the 'text.rst' file which updated 'Working with text data' for str dtype in pandas 3.0

I noticed that after updating the branch, a commit for expressions.py was added, although I have made no commits in the file. Could you please help me understand the changes and guide me through?

File: doc/source/user_guide/text.rst [hunk 1]
```diff
@@ -15,8 +15,9 @@ There are two ways to store text data in pandas:
 
 1. ``object`` -dtype NumPy array.
 2. :class:`StringDtype` extension type.
+3. ``str`` -dtype (default from pandas 3.0).
```
Comment (rhshadrach) [inline doc/source/user_guide/text.rst:18]: L14 needs to be updated. Also, should mention `string` here too I think.

File: doc/source/user_guide/text.rst [hunk 2]
```diff
@@ -15,8 +15,9 @@ There are two ways to store text data in pandas:
...
 
-We recommend using :class:`StringDtype` to store text data.
+We recommend using the ``str`` dtype or :class:`StringDtype` to store text data.
```
Comment (rhshadrach) [inline doc/source/user_guide/text.rst:20]: Or `string`. Maybe reword to something like

    We recommend __not__ using ``object`` dtype to store text data.

File: doc/source/user_guide/text.rst [hunk 3]
```diff
@@ -29,7 +30,15 @@ for many reasons:
...
    than ``'string'``.
 
-Currently, the performance of ``object`` dtype arrays of strings and
+Starting from pandas 3.0, the default dtype for string data is ``str``. Previously, ``object`` dtype was the
+default for string data. You may still encounter ``object`` dtype for legacy datasets or while working with earlier
+pandas versions. Use ``.astype("str")`` to explicitly convert these to ``str`` dtype or specify ``dtype="str"``
+when creating new data structures.
+
+Use the nullable :class:`StringDtype` (``"string"``) when handling NA values in your string data. It offers
+additional flexibility for missing values while maintaining compatibility with pandas' nullable types.
```
Comment (rhshadrach) [inline doc/source/user_guide/text.rst:39]: ``str`` also handles NA values, the difference between `np.NaN` vs `pd.NA`. I think it'd be good to clarify that here.

File: doc/source/user_guide/text.rst [hunk 4]
```diff
@@ -81,16 +92,42 @@ or convert from existing pandas data:
...
+   type(s3_string[0])  # pandas' StringDtype
 
 .. _text.differences:
 
 Behavior differences
 ^^^^^^^^^^^^^^^^^^^^
 
-These are places where the behavior of ``StringDtype`` objects differ from
-``object`` dtype
+These are places where the behavior of ``StringDtype`` or ``str`` objects differ from
```
Comment (rhshadrach) [inline doc/source/user_guide/text.rst:127]: Needs a mention of `string` too.

File: doc/source/user_guide/text.rst [hunk 5]
```diff
@@ -118,17 +155,15 @@ l. For ``StringDtype``, :ref:`string accessor methods<api.series.str>`
...
-   by a ``StringArray`` will return an object with :class:`BooleanDtype`,
-   rather than a ``bool`` dtype object. Missing values in a ``StringArray``
-   will propagate in comparison operations, rather than always comparing
-   unequal like :attr:`numpy.nan`.
+2. Some string methods, like :meth:`Series.str.decode`, are not available
+   on ``StringArray`` or ``str`` because they only hold strings, not bytes.
+3. In comparison operations, :class:`arrays.StringArray`, ``Series`` backed
+   by a ``StringArray``, and ``str`` dtype will return an object with :class:`BooleanDtype`,
+   rather than a ``bool`` dtype object. Missing values in these types will propagate
+   in comparison operations, rather than always comparing unequal like :attr:`numpy.nan`.
```
Comment (rhshadrach) [inline doc/source/user_guide/text.rst:163]: I don't think this is true for `str`.

File: doc/source/user_guide/text.rst [hunk 6]
```diff
@@ -187,27 +222,27 @@ and replacing any remaining whitespaces with underscores:
...
-    ``.str`` methods which operate on elements of type ``list`` are not available on such a
+    ``.str`` methods that operate on elements of type ``list`` are not available on such a
     ``Series``.
 
 .. _text.warn_types:
 
 .. warning::
 
-    The type of the Series is inferred and is one among the allowed types (i.e. strings).
+    The type of the Series is inferred as ``str`` or ``string`` depending on the context.
```
Comment (rhshadrach) [inline doc/source/user_guide/text.rst:242]: In what context do we infer ``string``?

File: doc/source/user_guide/text.rst [hunk 7]
```diff
@@ -396,7 +431,7 @@ Missing values on either side will result in missing values in the result as wel
 Concatenating a Series and something array-like into a Series
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-The parameter ``others`` can also be two-dimensional. In this case, the number or rows must match the lengths of the calling ``Series`` (or ``Index``).
+The parameter ``others`` can also be two-dimensional. In this case, the number or rows must match the length of the calling ``Series`` (or ``Index``).
```
Comment (rhshadrach) [inline doc/source/user_guide/text.rst:434]: `the number or rows` should be `of` instead

File: pandas/core/computation/expressions.py [hunk 8]
```diff
@@ -257,6 +257,12 @@ def where(cond, left_op, right_op, use_numexpr: bool = True):
         Whether to try to use numexpr.
     """
     assert _where is not None
+    
+    return (
+        _where(cond, left_op, right_op)
+        if use_numexpr
+        else _where_standard(cond, left_op, right_op)
+    )
```
Comment (rhshadrach) [inline pandas/core/computation/expressions.py:265]: This unintentional change was due to merging main, can you revert.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Thanks for the PR! I think it'd be good to also link to the PDEP for history - perhaps at the top?

https://pandas.pydata.org/pdeps/0014-string-dtype.html

File: doc/source/user_guide/text.rst (hunk 6 above)
Comment (Uvi-12) [reply doc/source/user_guide/text.rst:242]: The `string` dtype is inferred when the data includes `pd.NA` or other nullable types, ensuring compatibility with pandas' nullable ecosystem. It is also inferred when explicitly specified by the user with `dtype="string"`. Otherwise, `str` is typically inferred for text data. Please correct me if I am wrong.

**Comment by Uvi-12** [comment]
I have made changes according to the review, but somehow I can't revert the changes in expressions.py. Can you please help me with it?

**Comment by rhshadrach** [comment]
> I have made changes according to the review, but somehow I can't revert the changes in expressions.py. Can you please help me with it?

You can open the file with an editor and manually modify the file undoing the changes. Locally, you can run

    git diff upstream/main pandas/core/computation/expressions.py

and keep making modifications until that reports no differences.

**Comment by Uvi-12** [comment]
> If you'd like, I can push a commit undoing the changes there.

It would be very helpful, Thank you.

I have made the changes as per the review, please let me know if there are any other modifications required in the documentation.

File: doc/source/user_guide/text.rst (hunk 6 above)
Comment (rhshadrach) [reply doc/source/user_guide/text.rst:247]: > The `string` dtype is inferred when the data includes `pd.NA` or other nullable types, ensuring compatibility with pandas' nullable ecosystem.

pandas will not infer `string` here. This is consistent with other nullable dtypes.

```python
pd.set_option("future.infer_string", True)

ser = pd.Series(["a", pd.NA, "c"])
print(ser.dtype)
# str

ser = pd.Series([1, pd.NA, 3])
print(ser.dtype)
# object
```

> It is also inferred when explicitly specified by the user with `dtype="string"`.

This is not inference - inference, by definition, is the behavior when it is not explicitly specified.

Comment (Uvi-12) [reply doc/source/user_guide/text.rst:247]: Thank you for the clarification, I will change it back to the original. Are there any other modifications required?

**Comment by Uvi-12** [comment]
@rhshadrach, is everything alright, or are modifications needed?

File: doc/source/user_guide/text.rst (hunk 1 above)
Comment (rhshadrach) [reply doc/source/user_guide/text.rst:18]: The way this is written, it sounds like `str` is entirely separate from `StringDtype`, but that isn't the case.

```python
pd.set_option("future.infer_string", True)
ser1 = pd.Series(list("xyz"), dtype="str")
ser2 = pd.Series(list("xyz"), dtype=pd.StringDtype("pyarrow", np.nan))
print(ser1.dtype == ser2.dtype)
# True
```

Maybe just mention that `"str"`, when `future.infer_string` is set to True, is an alias for `pd.StringDtype("pyarrow", np.nan))` or `pd.StringDtype("python", np.nan))` depending on whether `pyarrow` is installed.

File: doc/source/user_guide/text.rst [hunk 9]
```diff
@@ -29,7 +34,16 @@ for many reasons:
...
    than ``'string'``.
 
-Currently, the performance of ``object`` dtype arrays of strings and
+Starting from pandas 3.0, the default dtype for string data is ``str``. Previously, ``object`` dtype was the
+default for string data. You may still encounter ``object`` dtype for legacy datasets or while working with earlier
+pandas versions. Use ``.astype("str")`` to explicitly convert these to ``str`` dtype or specify ``dtype="str"``
+when creating new data structures.
+
+Use the nullable :class:`StringDtype` (``"string"``) or ``str`` dtype when handling NA values in your string data.
+Note that ``StringDtype`` uses ``pd.NA`` for missing values, whereas ``str`` dtype uses ``np.NaN``. ``StringDtype``
```
Comment (rhshadrach) [inline doc/source/user_guide/text.rst:43]: This is incorrect, since `StringDtype` accepts an `na_value` argument to choose between `pd.NA` and `np.nan`. `np.NaN` has been removed as of NumPy 2.0, so shouldn't be used.

File: doc/source/user_guide/text.rst [hunk 10]
```diff
@@ -46,10 +60,11 @@ infer a list of strings to
...
    pd.Series(["a", "b", "c"])
 
-To explicitly request ``string`` dtype, specify the ``dtype``
+To explicitly request ``string`` or ``str`` dtype, specify the ``dtype``:
 
 .. ipython:: python
 
+   pd.Series(["a", "b", "c"], dtype="str")
    pd.Series(["a", "b", "c"], dtype="string")
    pd.Series(["a", "b", "c"], dtype=pd.StringDtype())
```
Comment (rhshadrach) [inline doc/source/user_guide/text.rst:69]: I think this should also show passing arguments to StringDtype.

File: doc/source/user_guide/text.rst (hunk 4 above)
Comment (rhshadrach) [reply doc/source/user_guide/text.rst:132]: Actually - can just leave this as ``StringDtype`` - no need to mention the aliases.

File: doc/source/user_guide/text.rst (hunk 6 above)
Comment (rhshadrach) [reply doc/source/user_guide/text.rst:247]: I think:

> When the option ``future.infer_string`` is set to True, the type of the Series is inferred as ``str``.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
There are several mentions of `` `pd.StringDtype` or `str` `` as if these are distinct entities. As mentioned below, they are not. I think we should make it clear that `str` is an alias, and just mention `pd.StringDtype` from there on. But open to other approaches too.

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.
